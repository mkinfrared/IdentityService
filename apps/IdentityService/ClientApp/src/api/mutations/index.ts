export * from "./loginMutation";

export * from "./registerMutation";

export * from "./verifyEmailMutation";

export * from "./submitConsent";

export * from "./submitForgotPassword";

export * from "./submitPasswordChange";
