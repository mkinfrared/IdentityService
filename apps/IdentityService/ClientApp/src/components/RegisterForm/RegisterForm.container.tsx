import { memo } from "react";

import { useReturnUrl } from "hooks";

import RegisterFormFC from "./RegisterForm";
import { RegisterFormContainerProps } from "./RegisterForm.type";

const RegisterFormContainer = (props: RegisterFormContainerProps) => {
  const returnUrl = useReturnUrl();

  return <RegisterFormFC {...props} returnUrl={returnUrl} />;
};

export default memo(RegisterFormContainer);
