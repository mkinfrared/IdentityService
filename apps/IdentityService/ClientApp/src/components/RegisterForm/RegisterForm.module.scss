@import "../../styles/common";

.RegisterForm {
  @extend %login-register-form;
}
