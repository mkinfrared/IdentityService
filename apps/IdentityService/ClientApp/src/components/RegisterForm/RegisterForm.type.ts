export type RegisterFormProps = {
  /**
   * a string that will be applied as a css class to parent element
   */
  className?: string;
  returnUrl: string;
};

export type RegisterFormContainerProps = Omit<RegisterFormProps, "returnUrl">;
