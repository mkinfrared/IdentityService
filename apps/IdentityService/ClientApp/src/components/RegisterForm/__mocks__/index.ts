export { default } from "./RegisterForm.mock";
