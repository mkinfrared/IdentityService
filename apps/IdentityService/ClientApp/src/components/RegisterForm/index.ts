export { default } from "./RegisterForm.container";

export * from "./RegisterForm.type";
