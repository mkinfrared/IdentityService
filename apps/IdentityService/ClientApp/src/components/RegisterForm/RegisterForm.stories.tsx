import { Meta, Story } from "@storybook/react";

import { withRouter } from "utils/storybookDecorators";

import { RegisterForm } from "./RegisterForm";
import { RegisterFormProps } from "./RegisterForm.type";
import css from "./Story.module.scss";

export default {
  title: "COMPONENTS/RegisterForm",
  component: RegisterForm,
  parameters: {
    componentSubtitle: "Subtitle goes here",
  },
  decorators: [withRouter],
} as Meta;

const Template: Story<RegisterFormProps> = (args) => (
  <div className={css.Story}>
    <RegisterForm {...args} />
  </div>
);

const Default = Template.bind({});

Default.args = {};

Default.parameters = {
  docs: {
    storyDescription: "Story description",
  },
};

export { Default };
