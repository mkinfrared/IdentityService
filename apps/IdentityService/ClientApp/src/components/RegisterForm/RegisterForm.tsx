import { yupResolver } from "@hookform/resolvers/yup";
import {
  classNames,
  getFormErrors,
  usePasswordVisibility,
} from "@identity-service/core";
import { Button, Heading, Text } from "@identity-service/ui";
import { memo } from "react";
import { useForm } from "react-hook-form";

import { registerMutation } from "api/mutations";
import { RegisterCommand } from "api/types";
import AppLink from "components/AppLink";
import BrandButton, { Brand } from "components/BrandButton";
import InputField from "components/InputField";
import { Routes } from "pages/Main/Main.type";
import { MainRoutes } from "routes/MainRoute";
import { isAxiosError } from "utils/api";
import { registerSchema } from "utils/validationSchemas";

import css from "./RegisterForm.module.scss";
import { RegisterFormProps } from "./RegisterForm.type";

const RegisterForm = ({ className, returnUrl }: RegisterFormProps) => {
  const defaultValues = {
    username: "",
    firstName: "",
    lastName: "",
    email: "",
    phoneNumber: "",
    password: "",
    passwordConfirmation: "",
    redirectUrl: returnUrl,
  };

  const { register, formState, handleSubmit, setError } =
    useForm<RegisterCommand>({
      defaultValues,
      mode: "onChange",
      resolver: yupResolver(registerSchema),
    });

  const { errors, isValid, isSubmitting } = formState;
  const { togglePasswordVisibility, fieldType, Icon } = usePasswordVisibility();

  const onSubmit = async (formData: RegisterCommand) => {
    try {
      const { token, userId } = await registerMutation(formData);

      sessionStorage.setItem("emailToken", token ?? "");

      sessionStorage.setItem("userId", userId);

      sessionStorage.setItem("returnUrl", returnUrl);

      const { location, history } = window;
      const pathname = `/account/${Routes.CONFIRM_EMAIL}`;
      const url = new URL(location.href);

      url.pathname = pathname;

      history.replaceState({}, "", url.toString());
    } catch (e) {
      if (isAxiosError(e)) {
        const { response } = e;

        const submitErrors = getFormErrors(
          response?.data as BadRequest<RegisterCommand>,
        );

        submitErrors.forEach(({ name, message, type }) => {
          setError(name, { message, type });
        });
      }

      console.error(e);
    }
  };

  const getExternalProviderLink = (brand: Brand) => {
    const url = new URL(`${window.location.origin}/auth/externalRegister`);

    const params = new URLSearchParams({
      provider: brand,
      returnUrl: encodeURIComponent(returnUrl),
    });

    url.search = params.toString();

    return url.toString();
  };

  return (
    <div className={classNames(css.RegisterForm, className)}>
      <form className={css.form} onSubmit={handleSubmit(onSubmit)}>
        <Heading className={css.heading}>Register</Heading>
        <InputField
          {...register("username")}
          className={css.input}
          label="Username"
          error={errors.username?.message}
        />
        <InputField
          {...register("firstName")}
          className={css.input}
          label="First Name"
          error={errors.firstName?.message}
        />
        <InputField
          {...register("lastName")}
          className={css.input}
          label="Last Name"
          error={errors.lastName?.message}
        />
        <InputField
          {...register("email")}
          className={css.input}
          label="Email"
          error={errors.email?.message}
        />
        <InputField
          {...register("phoneNumber")}
          className={css.input}
          label="Phone Number"
          error={errors.phoneNumber?.message}
        />
        <InputField
          {...register("password")}
          className={css.input}
          label="Password"
          suffix={Icon}
          onSuffixClick={togglePasswordVisibility}
          type={fieldType}
          error={errors.password?.message}
        />
        <InputField
          {...register("passwordConfirmation")}
          className={css.input}
          label="Confirm Password"
          suffix={Icon}
          onSuffixClick={togglePasswordVisibility}
          type={fieldType}
          error={errors.passwordConfirmation?.message}
        />
        <input
          type="text"
          value={returnUrl}
          hidden
          name="redirectUrl"
          readOnly
        />
        <Text className={css.text}>
          Already have an account. Consider{" "}
          <AppLink to={MainRoutes.LOGIN}>login</AppLink>
        </Text>
        <div className={css.buttonWrapper}>
          <Button
            disabled={!isValid || isSubmitting}
            type="submit"
            loading={isSubmitting}
          >
            Register
          </Button>
          <BrandButton
            href={getExternalProviderLink(Brand.GOOGLE)}
            brand={Brand.GOOGLE}
          >
            Register with Google
          </BrandButton>
        </div>
      </form>
    </div>
  );
};

export { RegisterForm };

export default memo(RegisterForm);
