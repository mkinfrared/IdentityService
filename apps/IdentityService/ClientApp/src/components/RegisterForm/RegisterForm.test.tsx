import { fireEvent, render, waitFor } from "@testing-library/react";

import * as mutations from "api/mutations";

import { RegisterForm } from "./RegisterForm";

jest.mock("api/mutations");

jest.mock("components/AppLink");

describe("<RegisterForm />", () => {
  const { location } = window;
  const mutationsMock = mutations as jest.Mocked<typeof mutations>;
  const returnUrl = "https://marklar.com";
  const Component = <RegisterForm returnUrl={returnUrl} />;

  beforeAll(() => {
    Object.defineProperty(window, "location", {
      configurable: true,
      value: {
        ...location,
        replace: jest.fn(),
      },
    });
  });

  afterAll(() => {
    Object.defineProperty(window, "location", {
      configurable: true,
      value: location,
    });
  });

  it("should be defined", () => {
    expect(RegisterForm).toBeDefined();
  });

  it("should match the snapshot", () => {
    const { container } = render(Component);

    expect(container).toMatchSnapshot();
  });

  it("should have a disabled button if fields are empty", () => {
    const { getByTestId } = render(Component);
    const submitButton = getByTestId("Button");

    expect(submitButton).toBeDisabled();
  });

  it("should have an enabled button if fields are filled and valid", async () => {
    const { getByTestId, container } = render(Component);
    const userNameField = container.querySelector('input[name="username"]');
    const firstNameField = container.querySelector('input[name="firstName"]');
    const lastNameField = container.querySelector('input[name="lastName"]');
    const emailField = container.querySelector('input[name="email"]');

    const phoneNumberField = container.querySelector(
      'input[name="phoneNumber"]',
    );

    const passwordField = container.querySelector('input[name="password"]');

    const passwordConfirmationField = container.querySelector(
      'input[name="passwordConfirmation"]',
    );

    const submitButton = getByTestId("Button");

    expect(submitButton).toBeDisabled();

    fireEvent.change(userNameField!, { target: { value: "foobar" } });

    fireEvent.change(firstNameField!, { target: { value: "foobar" } });

    fireEvent.change(lastNameField!, { target: { value: "foobar" } });

    fireEvent.change(emailField!, { target: { value: "foo@bar.com" } });

    fireEvent.change(phoneNumberField!, { target: { value: "foobar" } });

    fireEvent.change(passwordField!, { target: { value: "Marklar42$" } });

    fireEvent.change(passwordConfirmationField!, {
      target: { value: "Marklar42$" },
    });

    await waitFor(() => expect(submitButton).toBeEnabled());
  });

  it("should call registerMutation on submit", async () => {
    const userId = "00000000-0000-0000-0000-000000000000";
    const token = "fake-token";
    const registerResponse = { userId, token, code: null };
    const replaceState = jest.spyOn(window.history, "replaceState");

    mutationsMock.registerMutation.mockResolvedValueOnce(registerResponse);

    const data = {
      username: "marklar",
      firstName: "Timmy",
      lastName: "Burch",
      email: "timmy@coons.com",
      phoneNumber: "3334445566",
      password: "Marklar42$",
      passwordConfirmation: "Marklar42$",
      redirectUrl: returnUrl,
    };

    const { getByTestId, container } = render(Component);
    const userNameField = container.querySelector('input[name="username"]');
    const firstNameField = container.querySelector('input[name="firstName"]');
    const lastNameField = container.querySelector('input[name="lastName"]');
    const emailField = container.querySelector('input[name="email"]');

    const phoneNumberField = container.querySelector(
      'input[name="phoneNumber"]',
    );

    const passwordField = container.querySelector('input[name="password"]');

    const passwordConfirmationField = container.querySelector(
      'input[name="passwordConfirmation"]',
    );

    const submitButton = getByTestId("Button");

    fireEvent.change(userNameField!, { target: { value: data.username } });

    fireEvent.change(firstNameField!, { target: { value: data.firstName } });

    fireEvent.change(lastNameField!, { target: { value: data.lastName } });

    fireEvent.change(emailField!, { target: { value: data.email } });

    fireEvent.change(phoneNumberField!, {
      target: { value: data.phoneNumber },
    });

    fireEvent.change(passwordField!, { target: { value: data.password } });

    fireEvent.change(passwordConfirmationField!, {
      target: { value: data.passwordConfirmation },
    });

    await waitFor(() => expect(submitButton).toBeEnabled());

    await fireEvent.submit(submitButton);

    await waitFor(() =>
      expect(mutationsMock.registerMutation).toHaveBeenCalled(),
    );

    expect(mutationsMock.registerMutation).toHaveBeenCalledTimes(1);

    expect(mutationsMock.registerMutation).toHaveBeenCalledWith(data);

    expect(replaceState).toHaveBeenCalled();
  });

  it("should display an error on bad network request", async () => {
    const data = {
      username: "marklar",
      firstName: "Timmy",
      lastName: "Burch",
      email: "timmy@coons.com",
      phoneNumber: "3334445566",
      password: "Marklar42$",
      passwordConfirmation: "Marklar42$",
      redirectUrl: returnUrl,
    };

    const errorMessage = "Invalid login credentials";

    const axiosError = {
      isAxiosError: true,
      response: {
        data: {
          Username: ["Invalid login credentials"],
        },
        status: 400,
      },
    };

    mutationsMock.registerMutation.mockRejectedValueOnce(axiosError);

    const { getByTestId, container, getByText } = render(Component);
    const userNameField = container.querySelector('input[name="username"]');
    const firstNameField = container.querySelector('input[name="firstName"]');
    const lastNameField = container.querySelector('input[name="lastName"]');
    const emailField = container.querySelector('input[name="email"]');

    const phoneNumberField = container.querySelector(
      'input[name="phoneNumber"]',
    );

    const passwordField = container.querySelector('input[name="password"]');

    const passwordConfirmationField = container.querySelector(
      'input[name="passwordConfirmation"]',
    );

    const submitButton = getByTestId("Button");

    fireEvent.change(userNameField!, { target: { value: data.username } });

    fireEvent.change(firstNameField!, { target: { value: data.firstName } });

    fireEvent.change(lastNameField!, { target: { value: data.lastName } });

    fireEvent.change(emailField!, { target: { value: data.email } });

    fireEvent.change(phoneNumberField!, {
      target: { value: data.phoneNumber },
    });

    fireEvent.change(passwordField!, { target: { value: data.password } });

    fireEvent.change(passwordConfirmationField!, {
      target: { value: data.passwordConfirmation },
    });

    await waitFor(() => expect(submitButton).toBeEnabled());

    await fireEvent.submit(submitButton);

    await waitFor(() =>
      expect(mutationsMock.registerMutation).toHaveBeenCalled(),
    );

    expect(getByText(errorMessage)).toBeInTheDocument();
  });
});
