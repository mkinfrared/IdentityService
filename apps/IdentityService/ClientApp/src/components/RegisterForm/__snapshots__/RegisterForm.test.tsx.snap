// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegisterForm /> should match the snapshot 1`] = `
<div>
  <div
    class="RegisterForm"
  >
    <form
      class="form"
    >
      <h3
        class="Heading-module_Heading__Wg8yB heading"
        data-testid="Heading"
      >
        Register
      </h3>
      <label
        class="TextField-module_TextField__QtlJl input"
        data-testid="TextField"
      >
        <h5
          class="Heading-module_Heading__Wg8yB TextField-module_label__OD4N8"
          data-testid="Heading"
        >
          Username
        </h5>
        <div
          class="TextField-module_input__SPywc"
        >
          <input
            name="username"
            type="text"
            value=""
          />
        </div>
        <p
          class="Text-module_Text__K8mLZ TextField-module_error__IGQzp"
          data-testid="Text"
        />
      </label>
      <label
        class="TextField-module_TextField__QtlJl input"
        data-testid="TextField"
      >
        <h5
          class="Heading-module_Heading__Wg8yB TextField-module_label__OD4N8"
          data-testid="Heading"
        >
          First Name
        </h5>
        <div
          class="TextField-module_input__SPywc"
        >
          <input
            name="firstName"
            type="text"
            value=""
          />
        </div>
        <p
          class="Text-module_Text__K8mLZ TextField-module_error__IGQzp"
          data-testid="Text"
        />
      </label>
      <label
        class="TextField-module_TextField__QtlJl input"
        data-testid="TextField"
      >
        <h5
          class="Heading-module_Heading__Wg8yB TextField-module_label__OD4N8"
          data-testid="Heading"
        >
          Last Name
        </h5>
        <div
          class="TextField-module_input__SPywc"
        >
          <input
            name="lastName"
            type="text"
            value=""
          />
        </div>
        <p
          class="Text-module_Text__K8mLZ TextField-module_error__IGQzp"
          data-testid="Text"
        />
      </label>
      <label
        class="TextField-module_TextField__QtlJl input"
        data-testid="TextField"
      >
        <h5
          class="Heading-module_Heading__Wg8yB TextField-module_label__OD4N8"
          data-testid="Heading"
        >
          Email
        </h5>
        <div
          class="TextField-module_input__SPywc"
        >
          <input
            name="email"
            type="text"
            value=""
          />
        </div>
        <p
          class="Text-module_Text__K8mLZ TextField-module_error__IGQzp"
          data-testid="Text"
        />
      </label>
      <label
        class="TextField-module_TextField__QtlJl input"
        data-testid="TextField"
      >
        <h5
          class="Heading-module_Heading__Wg8yB TextField-module_label__OD4N8"
          data-testid="Heading"
        >
          Phone Number
        </h5>
        <div
          class="TextField-module_input__SPywc"
        >
          <input
            name="phoneNumber"
            type="text"
            value=""
          />
        </div>
        <p
          class="Text-module_Text__K8mLZ TextField-module_error__IGQzp"
          data-testid="Text"
        />
      </label>
      <label
        class="TextField-module_TextField__QtlJl input TextField-module_hasSuffix__-bHtc"
        data-testid="TextField"
      >
        <h5
          class="Heading-module_Heading__Wg8yB TextField-module_label__OD4N8"
          data-testid="Heading"
        >
          Password
        </h5>
        <div
          class="TextField-module_input__SPywc"
        >
          <input
            name="password"
            type="password"
            value=""
          />
          <div
            class="FakeButton-module_FakeButton__-W3X5 TextField-module_suffix__SCuij"
            data-testid="FakeButton"
            role="button"
            tabindex="0"
          >
            <svg
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
              />
            </svg>
          </div>
        </div>
        <p
          class="Text-module_Text__K8mLZ TextField-module_error__IGQzp"
          data-testid="Text"
        />
      </label>
      <label
        class="TextField-module_TextField__QtlJl input TextField-module_hasSuffix__-bHtc"
        data-testid="TextField"
      >
        <h5
          class="Heading-module_Heading__Wg8yB TextField-module_label__OD4N8"
          data-testid="Heading"
        >
          Confirm Password
        </h5>
        <div
          class="TextField-module_input__SPywc"
        >
          <input
            name="passwordConfirmation"
            type="password"
            value=""
          />
          <div
            class="FakeButton-module_FakeButton__-W3X5 TextField-module_suffix__SCuij"
            data-testid="FakeButton"
            role="button"
            tabindex="0"
          >
            <svg
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
              />
            </svg>
          </div>
        </div>
        <p
          class="Text-module_Text__K8mLZ TextField-module_error__IGQzp"
          data-testid="Text"
        />
      </label>
      <input
        hidden=""
        name="redirectUrl"
        readonly=""
        type="text"
        value="https://marklar.com"
      />
      <p
        class="Text-module_Text__K8mLZ text"
        data-testid="Text"
      >
        Already have an account. Consider
         
        <div
          data-testid="AppLink"
        >
          {"to":"/account/login","children":"login"}
        </div>
      </p>
      <div
        class="buttonWrapper"
      >
        <button
          class="Button-module_Button__0OJEu Button-module_opaque__O0DEF"
          data-testid="Button"
          disabled=""
          type="submit"
        >
          <svg
            class="Button-module_spinner__kVZYq"
            fill="#fff"
            viewBox="0 0 135 140"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="120"
              rx="6"
              width="15"
              y="10"
            >
              <animate
                attributeName="height"
                begin="0.5s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="120;110;100;90;80;70;60;50;40;140;120"
              />
              <animate
                attributeName="y"
                begin="0.5s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="10;15;20;25;30;35;40;45;50;0;10"
              />
            </rect>
            <rect
              height="120"
              rx="6"
              width="15"
              x="30"
              y="10"
            >
              <animate
                attributeName="height"
                begin="0.25s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="120;110;100;90;80;70;60;50;40;140;120"
              />
              <animate
                attributeName="y"
                begin="0.25s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="10;15;20;25;30;35;40;45;50;0;10"
              />
            </rect>
            <rect
              height="140"
              rx="6"
              width="15"
              x="60"
            >
              <animate
                attributeName="height"
                begin="0s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="120;110;100;90;80;70;60;50;40;140;120"
              />
              <animate
                attributeName="y"
                begin="0s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="10;15;20;25;30;35;40;45;50;0;10"
              />
            </rect>
            <rect
              height="120"
              rx="6"
              width="15"
              x="90"
              y="10"
            >
              <animate
                attributeName="height"
                begin="0.25s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="120;110;100;90;80;70;60;50;40;140;120"
              />
              <animate
                attributeName="y"
                begin="0.25s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="10;15;20;25;30;35;40;45;50;0;10"
              />
            </rect>
            <rect
              height="120"
              rx="6"
              width="15"
              x="120"
              y="10"
            >
              <animate
                attributeName="height"
                begin="0.5s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="120;110;100;90;80;70;60;50;40;140;120"
              />
              <animate
                attributeName="y"
                begin="0.5s"
                calcMode="linear"
                dur="1s"
                repeatCount="indefinite"
                values="10;15;20;25;30;35;40;45;50;0;10"
              />
            </rect>
          </svg>
          <span
            class="Button-module_content__PpYVU"
          >
            Register
          </span>
        </button>
        <a
          class="LinkButton-module_LinkButton__x1Ppn BrandButton google LinkButton-module_opaque__DFeur"
          data-testid="LinkButton"
          href="http://localhost/auth/externalRegister?provider=Google&returnUrl=https%253A%252F%252Fmarklar.com"
        >
          <svg
            class="brandIcon"
          >
            google.svg
          </svg>
          Register with Google
        </a>
      </div>
    </form>
  </div>
</div>
`;
